sw/device/tests/rstmgr_sw_rst_ctrl_test.c

```diff
@@ -0,0 +1,327 @@
+ * // 2     | SPI_HOST1  |  CONFIGOPTS    |  0x0         |  0x6540000
+ * // 3     | USB        |  EP_OUT_ENABLE |  0x0         |  0xc3
+ * // 4     | USBIF      |  RXENABLE_OUT  |  0x0         |  0x695
+ * // 5     | I2C0       |  TIMING0       |  0x0         |  0x8b00cfe
+ * // 6     | I2C1       |  TIMING1       |  0x0         |  0x114010d8
+ * // 7     | I2C2       |  TIMING2       |  0x0         |  0x19ec1595
+ *
+ * 'test register' is a rw type register under each peripheral device.
+ * During the test, these registers are programmed with arbitrary values. ('prgm
+ * value') when sw resets are asserted, those values are updated to reset value.
+ * In each round, the rstmgr asserts reset to each peripheral device and
+ * each test register value is compared with the expected reset value.
+ */
+
+#define kRstMgr kTopEarlgreyResetManagerSwResetsLast + 1
+#define kNumNodes kRstMgr + 1
+
+typedef struct node node_t;
+typedef void config_f(node_t *);
+
+// peripheral device template
+// name : device name
+// base : base address
+// offset : offset to the test register
+// handler : dif handler for each peripheral device
+// config_f : function ptr to each peripheral program
+// reset_val : reset value for comparison
+// prgm_val : expected value for test register after configuration
+//            function is called.
+struct node {
+  const char *name;
+  uintptr_t base;
+  ptrdiff_t offset;
+  void * handler;
+  config_f *config;
+  uint32_t reset_val;
+  uint32_t prgm_val;
+};
+
+// handler declaration
+dif_rstmgr_t rstmgr;
+dif_spi_device_t spi_dev;
+dif_spi_host_t spi_host0, spi_host1;
+dif_usbdev_t usbdev;
+dif_i2c_t i2c0, i2c1, i2c2;
+
+// pre declaration of template
+static node_t test_node[kNumNodes];
+
+// test register read from a given peripheral device
+uint32_t read_test_reg(node_t *node) {
+  return (mmio_region_read32(mmio_region_from_addr(node->base), node->offset));
+}
+
+// init function
+void init_nodes(int n) {
+#define CALL_INIT(module_)      \
+  CHECK_DIF_OK(dif_##module_##_init( \
+      mmio_region_from_addr(test_node[n].base), (dif_##module_##_t *)test_node[n].handler));\
+  break  // No semicolon to force semicolon below.
+
+  switch (n) {
+    case kTopEarlgreyResetManagerSwResetsSpiDevice:
+      CALL_INIT(spi_device);
+    case kTopEarlgreyResetManagerSwResetsSpiHost0:
+      CALL_INIT(spi_host);
+    case kTopEarlgreyResetManagerSwResetsSpiHost1:
+      CALL_INIT(spi_host);
+    case kTopEarlgreyResetManagerSwResetsUsb:
+      CALL_INIT(usbdev);
+    // USB init covers USBIF too.
+    case kTopEarlgreyResetManagerSwResetsUsbif:
+      break;
+    case kTopEarlgreyResetManagerSwResetsI2c0:
+      CALL_INIT(i2c);
+    case kTopEarlgreyResetManagerSwResetsI2c1:
+      CALL_INIT(i2c);
+    case kTopEarlgreyResetManagerSwResetsI2c2:
+      CALL_INIT(i2c);
+    case kRstMgr:
+      CALL_INIT(rstmgr);
+    default:
+      LOG_FATAL("Unknown node %d", n);
```

Comment:
I will suggest unrolling the loop on line 298 here itself, rather than have this switch-case. So this will just look like this:
```c
  INITIALIZE(spi_device, kTopEarlgreyResetManagerSwResetsSpiDevice);
  INITIALIZE(spi_host, kTopEarlgreyResetManagerSwResetsSpiHost0);
  ...
```

You can pull the `reset_val` read into the macro.

Reply:
Updated. This looks a lot better. thx!